### PR TITLE
feat(lang): implement expose func

### PR DESCRIPTION
**Pull request by aseaday**
Signed-off-by: nullday <aseaday@hotmail.com>

### CODE REVIEW

**Comment by aseaday** [comment]
The following work will be:
- add describe https://github.com/tensorchord/envd/issues/782
- add test https://github.com/tensorchord/envd/issues/727

**Review by gaocegege — APPROVED** [review]
cc @kemingy

**Comment by aseaday** [comment]
> Is service used in the code?

The Defaultgraph save the service and it should be used in `envd envs describe` functions or future commands to shows multi service for a container/env

**Comment by gaocegege** [comment]
SGTM

File: pkg/lang/frontend/starlark/universe/const.go [hunk 1]
```diff
@@ -19,4 +19,5 @@ const (
 	ruleShell     = "shell"
 	ruleRun       = "run"
 	ruleGitConfig = "git_config"
+	ruleExpose    = "expose"
```
Comment (kemingy) [inline pkg/lang/frontend/starlark/universe/const.go:22]: * https://github.com/tensorchord/envd/pull/769#pullrequestreview-1072455645
* Will it be better to place `expose` under `runtime`?

Comment (aseaday) [reply pkg/lang/frontend/starlark/universe/const.go:22]: LGTM

File: pkg/lang/frontend/starlark/runtime/runtime.go [hunk 2]
```diff
@@ -90,3 +92,30 @@ func ruleFuncDaemon(thread *starlark.Thread, _ *starlark.Builtin,
...
+		hostPort    = starlark.MakeInt(0)
+		serviceName = starlark.String("")
+	)
+
+	if err := starlark.UnpackArgs(ruleExpose,
+		args, kwargs, "envd_port", &envdPort, "host_port?", &hostPort, "service?", &serviceName); err != nil {
+		return nil, err
+	}
+	envdPortInt, ok := envdPort.Int64()
+	if !ok && envdPortInt < 0 && envdPortInt > 65536 {
```
Comment (zwpaper) [inline pkg/lang/frontend/starlark/runtime/runtime.go:109]: here should be `||`?

File: pkg/lang/frontend/starlark/runtime/runtime.go [hunk 3]
```diff
@@ -90,3 +92,30 @@ func ruleFuncDaemon(thread *starlark.Thread, _ *starlark.Builtin,
...
+		return nil, errors.New("envd_port must be a positive integer less than 65536")
+	}
+	hostPortInt, ok := hostPort.Int64()
+	if !ok && hostPortInt < 0 && hostPortInt > 65536 {
```
Comment (zwpaper) [inline pkg/lang/frontend/starlark/runtime/runtime.go:113]: also should be `||`?

File: pkg/lang/frontend/starlark/runtime/runtime.go (hunk 2 above)
Comment (zwpaper) [inline pkg/lang/frontend/starlark/runtime/runtime.go:109]: `65536` can not be used to bind, `0` also may be reserved.

File: pkg/lang/frontend/starlark/runtime/runtime.go (hunk 3 above)
Comment (zwpaper) [inline pkg/lang/frontend/starlark/runtime/runtime.go:113]: `65536` can not be used to bind, `0` also may be reserved.

**Review by zwpaper — CHANGES_REQUESTED** [review]

File: pkg/lang/frontend/starlark/runtime/runtime.go (hunk 2 above)
Comment (aseaday) [reply pkg/lang/frontend/starlark/runtime/runtime.go]: Thx! I dreamed this bug too! LOL

File: pkg/lang/frontend/starlark/runtime/runtime.go (hunk 3 above)
Comment (aseaday) [reply pkg/lang/frontend/starlark/runtime/runtime.go]: Invited you as the co-author of this commit.

**Review by kemingy — APPROVED** [review]

**Review by zwpaper — APPROVED** [review]

**Pull request merged**